src/interrupt/mod.rs

```diff
@@ -0,0 +1,208 @@
+#[derive(Copy, Clone, Debug)]
+pub struct LegacyIrqGroupConfig {
+    /// Legacy irq number.
+    pub irq: InterruptIndex,
+}
+
+/// Configuration data for MSI/MSI-X interrupt groups
+///
+/// MSI/MSI-X interrupt groups are basically a set of vectors.
+#[derive(Copy, Clone, Debug)]
+pub struct MsiIrqGroupConfig {
+    /// First index of the MSI/MSI-X interrupt vectors
+    pub base: InterruptIndex,
+    /// Number of vectors in the MSI/MSI-X group.
+    pub count: InterruptIndex,
+}
+
+/// Trait to manage interrupt sources for virtual device backends.
+///
+/// The InterruptManager implementations should protect itself from concurrent accesses internally,
+/// so it could be invoked from multi-threaded context.
+pub trait InterruptManager {
+    type GroupConfig;
+
+    /// Create an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object to manage
+    /// interrupt sources for a virtual device
+    ///
+    /// An [InterruptSourceGroup](trait.InterruptSourceGroup.html) object manages all interrupt
+    /// sources of the same type for a virtual device.
+    ///
+    /// # Arguments
+    /// * config: The interrupt group configuration
+    fn create_group(&self, config: Self::GroupConfig)
+        -> Result<Arc<Box<dyn InterruptSourceGroup>>>;
+
+    /// Destroy an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object created by
+    /// [create_group()](trait.InterruptManager.html#tymethod.create_group).
+    ///
+    /// Assume the caller takes the responsibility to disable all interrupt sources of the group
+    /// before calling destroy_group(). This assumption helps to simplify InterruptSourceGroup
+    /// implementations.
+    fn destroy_group(&mut self, group: Arc<Box<dyn InterruptSourceGroup>>) -> Result<()>;
+}
+
+/// Configuration data for legacy interrupts.
+///
+/// On x86 platforms, legacy interrupts means those interrupts routed through PICs or IOAPICs.
+#[derive(Copy, Clone, Debug)]
+pub struct LegacyIrqSourceConfig {}
+
+/// Configuration data for MSI/MSI-X interrupts.
+///
+/// On x86 platforms, these interrupts are vectors delivered directly to the LAPIC.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct MsiIrqSourceConfig {
+    /// High address to delivery message signaled interrupt.
+    pub high_addr: u32,
+    /// Low address to delivery message signaled interrupt.
+    pub low_addr: u32,
+    /// Data to write to delivery message signaled interrupt.
+    pub data: u32,
+}
+
+/// Configuration data for an interrupt source.
+#[derive(Copy, Clone, Debug)]
+pub enum InterruptSourceConfig {
+    /// Configuration data for Legacy interrupts.
+    LegacyIrq(LegacyIrqSourceConfig),
+    /// Configuration data for PciMsi, PciMsix and generic MSI interrupts.
+    MsiIrq(MsiIrqSourceConfig),
+}
+
+pub trait InterruptSourceGroup: Send + Sync {
+    /// Enable the interrupt sources in the group to generate interrupts.
+    fn enable(&mut self) -> Result<()> {
```

Comment:
Suggest to revert enable() to:
fn enable(&self, configs: &[InterruptSourceConfig]) -> Result<()>;

The proposed work flow for virtio devices is:
1) guest kernel writes MSI configuration registers, the vmm caches those configuration information.
2) guest kernel writes virtio status register to start virtio device
2.1) the vmm applies cached msi configuration as a batch and commit them to kvm routing table. 
2.2) invoke virtio device's activate()
3) guest kernel writes MSI registers to update MSI configuration. Because the virtio device has been activated, the vmm calls InterruptSourceGroup::update() to directly update kvm routing table.

So we need to pass the cached MSI configuration information when calling  InterruptSourceGroup::enable().

And as Paolo suggested, it would be better for enable()/disable() to take "&self" instead of "&mut self". BTW, what's the scenario needing a "&mut self" for enable()/disable()?